Compansation dampening functionality is added

ahmetipkin opened the pull request:
Now the compensation is not applied always - rather is applied only for limited amount (like Repetier firmware does). In order to configure it, you add two new configs:
leveling-strategy.rectangular-grid.height_limit <height_limit>
leveling-strategy.rectangular-grid.dampening_start <start limit>

Basically what it does is; until Z axis is below 'start limit', any compensation that would be applied is applied as is (full power). However, after this 'start limit' is hit, until 'height limit' is reached, the compensation is dampened gradually (linearly) from full power to none (i.e. no compensation after 'height limit' is passed)

Main reason for this is: we generally use Mesh bed leveling for compensating our non-flat beds; but generally after first few layers; the extruded plastic can cover for this bent; and it is no longer needed.

ahmetipkin committed:
Added compensation degragation function for rectangular grid

Now the compensation is not applied always - rather is applied only for limited amount (like Repetier firmware does). In order to configure it, you add two new configs:
leveling-strategy.rectangular-grid.height_limit <height_limit>
leveling-strategy.rectangular-grid.dampening_start <start limit>

Basically what it does is; until Z axis is below 'start limit', any compensation that would be applied is applied as is (full power). However, after this 'start limit' is hit, until 'height limit' is reached, the compensation is dampened gradually (linearly) from full power to none (i.e. no compensation after 'height limit' is passed)

Main reason for this is: we generally use Mesh bed leveling for compensating our non-flat beds; but generally after first few layers; the extruded plastic can cover for this bent; and it is no longer needed.

wolfmanjm reviewed (changes requested):
This seems to always be active, and the default is set to 1.0mm, which means this is not a backward compatible change so is rejected.
For PRs to be accepted they need to have new features disabled by default, this one does not.

The relevant section in the PR submission guidelines is here.... 
"Maintain backwards compatibility. Do not make changes that will break other peoples systems, it is not acceptable to force everyone to change their config files to accommodate your change. If necessary add new configs to enable your change if it is breaking backward compatibility."

arthurwolf commented:
Thanks, this is a really nice feature, and I hope you can make it optional
so it can be considered for merging.

On Thu, Mar 1, 2018 at 3:14 AM, ahmetipkin <notifications@github.com> wrote:

> Now the compensation is not applied always - rather is applied only for
> limited amount (like Repetier firmware does). In order to configure it, you
> add two new configs:
> leveling-strategy.rectangular-grid.height_limit <height_limit>
> leveling-strategy.rectangular-grid.dampening_start
>
> Basically what it does is; until Z axis is below 'start limit', any
> compensation that would be applied is applied as is (full power). However,
> after this 'start limit' is hit, until 'height limit' is reached, the
> compensation is dampened gradually (linearly) from full power to none (i.e.
> no compensation after 'height limit' is passed)
>
> Main reason for this is: we generally use Mesh bed leveling for
> compensating our non-flat beds; but generally after first few layers; the
> extruded plastic can cover for this bent; and it is no longer needed.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Smoothieware/Smoothieware/pull/1274
> Commit Summary
>
>    - Added compensation degragation function for rectangular grid
>
> File Changes
>
>    - *M* src/modules/tools/zprobe/CartGridStrategy.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1274/files#diff-0>
>    (93)
>    - *M* src/modules/tools/zprobe/CartGridStrategy.h
>    <https://github.com/Smoothieware/Smoothieware/pull/1274/files#diff-1>
>    (8)
>
> Patch Links:
>
>    - https://github.com/Smoothieware/Smoothieware/pull/1274.patch
>    - https://github.com/Smoothieware/Smoothieware/pull/1274.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1274>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFcsWZeAGhtqcYyNfw6cHOYEJHLIXks5tZ1mRgaJpZM4SXpzY>
> .
>



-- 
Courage et bonne humeur.

ahmetipkin commented:
Makes sense. I can add a "enable_dampening" config - which defaults to false- and check it in the code accordingly; so this would cover that case.

Sorry, it was 3 AM when I made this push; I've missed that part in Guideline :)

wolfmanjm commented:
to save ram it is best to set the default to NAN, and then check for NAN. that way there is no need to add even more memory use

ahmetipkin commented:
Ok, will try that way. Not very comfortable with NaN's here though :)

Commit:
Made the Compansation Dampening as optional

If the config values are not set; the compansation dampening is shut down; so the compansation is applied everywhere - like the old behaviour

If you want to enable it; you should define both of

leveling-strategy.rectangular-grid.height_limit
leveling-strategy.rectangular-grid.dampening_start

configs.

ahmetipkin commented:
Ok, the change is in place now :)

wolfmanjm commented:
You committed three bogus files that are not allowed in a pull request. Please remove them. Thank you.

Commit:
Deleted the bogus files.

ahmetipkin commented:
Oh damn, right away! Sorry :)

wolfmanjm commented on src/modules/tools/zprobe/CartGridStrategy.cpp:
```diff
@@ -133,7 +135,15 @@ bool CartGridStrategy::handleConfig()
+    this->height_limit = THEKERNEL->config->value(leveling_strategy_checksum, cart_grid_leveling_strategy_checksum, height_limit_checksum)->by_default(NAN)->as_number();
+    this->dampening_start = THEKERNEL->config->value(leveling_strategy_checksum, cart_grid_leveling_strategy_checksum, dampening_start_checksum)->by_default(NAN)->as_number();
+
+    if(!isnan(this->height_limit) && !isnan(this->dampening_start))
```
it is bad practice to not use braces here. Although it works it is considered bad form.

wolfmanjm commented on src/modules/tools/zprobe/CartGridStrategy.cpp:
```diff
@@ -541,56 +551,73 @@ bool CartGridStrategy::doProbe(Gcode *gc)
+			if( current_z >= this->dampening_start)
+				scale = ( 1- ( (current_z - this->dampening_start ) / this->damping_interval) );
+			// else leave scale at 1.0;
+		} else
```
braces please. This is part of the coding guidelines I believe.

wolfmanjm commented on src/modules/tools/zprobe/CartGridStrategy.cpp:
```diff
@@ -541,56 +551,73 @@ bool CartGridStrategy::doProbe(Gcode *gc)
+		// if the height is below our compensation limit:
+		if(current_z <= this->height_limit) {
+			// scale the offset as necessary:
+			if( current_z >= this->dampening_start)
```
braces needed

wolfmanjm commented on src/modules/tools/zprobe/CartGridStrategy.cpp:
```diff
@@ -541,56 +551,73 @@ bool CartGridStrategy::doProbe(Gcode *gc)
+			scale = 0.0; // if Z is higher than max, no compensation
+	}
+
+	if(inverse)
```
please add braces, as per coding guidelines.

wolfmanjm commented on src/modules/tools/zprobe/CartGridStrategy.cpp:
```diff
@@ -541,56 +551,73 @@ bool CartGridStrategy::doProbe(Gcode *gc)
-        THEKERNEL->streams->printf("//DEBUG: right= %f\n", right);
-        THEKERNEL->streams->printf("//DEBUG: offset= %f\n", offset);
-    */
+	
```
Indentation is incorrect on this entire block, again coding guidelines stipulate 4 spaces per indent. thanks.

ahmetipkin replied:
Eclipse messed up apparently. Fixing.

ahmetipkin commented:
Thanks for the comments. On them right now.

Commit:
Fixed indentation and braces

wolfmanjm commented:
thank you that is perfect now.

ahmetipkin commented:
Thank you, it was a pleasure! 👍

arthurwolf commented:
This also needs to be added to the wiki documentation, would you mind doing
that ? The zprobe page would be the correct place. Make sure you also edit
the options list.

On Thu, Mar 1, 2018 at 5:29 PM, Ahmet Ipkin <notifications@github.com>
wrote:

> Thank you, it was a pleasure! 👍
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1274#issuecomment-369648446>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFfPdvbM1qjI8yS9ApcQ-0b1LqOiHks5taCHggaJpZM4SXpzY>
> .
>



-- 
Courage et bonne humeur.

ahmetipkin commented:
Sure thing. I'll add them now

ahmetipkin commented:
Done. 👍  Please review it :)

arthurwolf commented:
Just seen it, perfect, thank you very much !

On Thu, Mar 1, 2018 at 11:07 PM, Ahmet Ipkin <notifications@github.com>
wrote:

> Done. 👍 Please review it :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1274#issuecomment-369747785>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFbQVFEn7am2-rd-2Hh_RrPv7UnYyks5taHEbgaJpZM4SXpzY>
> .
>



-- 
Courage et bonne humeur.
